### PR TITLE
Make `torch.optim.RMSprop` differentiable

### PR DESCRIPTION
**Pull request by emcastillo**
Blocked by #82205

### CODE REVIEW

File: torch/optim/rmsprop.py [hunk 1]
```diff
@@ -125,20 +127,22 @@ def step(self, closure=None):
 
                 # State initialization
                 if len(state) == 0:
-                    state['step'] = 0
+                    state['step'] = torch.tensor(0)
```
Comment (albanD) [inline torch/optim/rmsprop.py:130]: I am not sure this is ok?
In particular when working on GPU, it makes a big difference if this is a Tensor or a scalar.

File: torch/optim/rmsprop.py [hunk 2]
```diff
@@ -232,9 +240,14 @@ def _single_tensor_rmsprop(params: List[Tensor],
...
             grad_avg = grad_avgs[i]
             grad_avg.mul_(alpha).add_(grad, alpha=1 - alpha)
-            avg = square_avg.addcmul(grad_avg, grad_avg, value=-1).sqrt_().add_(eps)
+            avg = square_avg.addcmul(grad_avg, grad_avg, value=-1).sqrt_()
         else:
-            avg = square_avg.sqrt().add_(eps)
+            avg = square_avg.sqrt()
+
+        if differentiable:
+            avg = avg.clone().add_(eps)
```
Comment (albanD) [inline torch/optim/rmsprop.py:248]: nit: just do an out of place add: `avg.add(eps)` ?

**Review by albanD — COMMENTED** [review]
Sounds good.
Do update the setstate similar to https://github.com/pytorch/pytorch/pull/83588

File: torch/optim/rmsprop.py (hunk 1 above)
Comment (vadimkantorov) [reply torch/optim/rmsprop.py:130]: I think earlier in some other PR having this as tensor was a problem... (as it gets put on another device if checkpoint is loaded directly onto GPU, then incrementing a GPU tensor becomes slow) I remember it being reverted somewhere...

File: torch/optim/rmsprop.py (hunk 2 above)
Comment (emcastillo) [reply torch/optim/rmsprop.py:248]: yeah, right, i was in autopilot mode ...

**Review by albanD — APPROVED** [review]
SGTM!

**Comment by albanD** [comment]
The AOTAutograd test was broken on master and a rebase should fix it.

**Pull request closed without merging**